src/testing/deploy/upgrade_test.go

```diff
@@ -76,16 +77,65 @@ func TestUpgradeTrigger(t *testing.T) {
 		t.Skip("Skipping upgrade test")
 	}
 	fromVersions := []string{
-		"2.4.6",
-		"2.5.0",
+		"2.7.6",
+		"2.8.5",
 	}
+
+	type ExpectedCommitCount struct {
+		preTrigger1  int
+		preTrigger2  int
+		postTrigger1 int
+		postTrigger2 int
+	}
+
+	getExpectedCommitCountFromVersion := func(version string) ExpectedCommitCount {
+		expectedCommitMap := map[string]ExpectedCommitCount{
+			"v2.7": {
+				// 2.7 is a special case where the commit structure changes after the upgrade
+				preTrigger1:  23,
+				preTrigger2:  12,
+				postTrigger1: 33,
+				postTrigger2: 17,
+			},
+			"v2.8": {
+				// trigger 1 is two empty commits and 11 data commits
+				// trigger 2 is one inital commit and "every other" data commit, so 6 total
+				preTrigger1:  13,
+				preTrigger2:  6,
+				postTrigger1: 13,
+				postTrigger2: 6,
+			},
+		}
+		lookup := semver.MajorMinor("v" + version)
+		return expectedCommitMap[lookup]
+	}
+
 	dataRepo := "TestTrigger_data"
 	dataCommit := client.NewCommit(pfs.DefaultProjectName, dataRepo, "master", "")
+	pipeline1 := "TestTrigger1"
+	pipeline2 := "TestTrigger2"
+
+	logCommits := func(t *testing.T, c *client.APIClient, commits []*pfs.CommitInfo) {
+		var buf bytes.Buffer
+		for i, commit := range commits {
+			err := c.GetFile(commit.Commit, "/hello", &buf)
+			commitFile := buf.String()
+			buf.Reset()
+			if err != nil || commitFile == "" {
```

Comment:
I'd check err before accessing anything on `buf`. If there's an error you can't really expect anything of the output. It could be a nil pointer here

Reply:
buf is always ready to use regardless of the outcome of GetFile, but it will obviously be annoying when the output says the file is empty, when in reality there was some error.